File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/CheckConfig.java
package biz.kryukov.dev.dephealth;

import java.time.Duration;

/**
 * Dependency health check configuration. Immutable, created via Builder.
 */
public final class CheckConfig {

    /** Default check interval: 15 seconds. */
    public static final Duration DEFAULT_INTERVAL = Duration.ofSeconds(15);
    /** Default check timeout: 5 seconds. */
    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(5);
    /** Default initial delay before first check: 5 seconds. */
    public static final Duration DEFAULT_INITIAL_DELAY = Duration.ofSeconds(5);
    /** Default failure threshold: 1 consecutive failure. */
    public static final int DEFAULT_FAILURE_THRESHOLD = 1;
    /** Default success threshold: 1 consecutive success. */
    public static final int DEFAULT_SUCCESS_THRESHOLD = 1;

    /** Minimum allowed check interval. */
    public static final Duration MIN_INTERVAL = Duration.ofSeconds(1);
    /** Maximum allowed check interval. */
    public static final Duration MAX_INTERVAL = Duration.ofMinutes(10);
    /** Minimum allowed check timeout. */
    public static final Duration MIN_TIMEOUT = Duration.ofMillis(100);
    /** Maximum allowed check timeout. */
    public static final Duration MAX_TIMEOUT = Duration.ofSeconds(30);
    /** Minimum allowed initial delay. */
    public static final Duration MIN_INITIAL_DELAY = Duration.ZERO;
    /** Maximum allowed initial delay. */
    public static final Duration MAX_INITIAL_DELAY = Duration.ofMinutes(5);
    /** Minimum allowed threshold value. */
    public static final int MIN_THRESHOLD = 1;
    /** Maximum allowed threshold value. */
    public static final int MAX_THRESHOLD = 10;

    private final Duration interval;
    private final Duration timeout;
    private final Duration initialDelay;
    private final int failureThreshold;
    private final int successThreshold;

    private CheckConfig(Builder builder) {
        this.interval = builder.interval;
        this.timeout = builder.timeout;
        this.initialDelay = builder.initialDelay;
        this.failureThreshold = builder.failureThreshold;
        this.successThreshold = builder.successThreshold;
    }

    /** Returns the check interval. */
    public Duration interval() {
        return interval;
    }

    /** Returns the check timeout. */
    public Duration timeout() {
        return timeout;
    }

    /** Returns the initial delay before the first check. */
    public Duration initialDelay() {
        return initialDelay;
    }

    /** Returns the number of consecutive failures to become unhealthy. */
    public int failureThreshold() {
        return failureThreshold;
    }

    /** Returns the number of consecutive successes to become healthy. */
    public int successThreshold() {
        return successThreshold;
    }

    /** Creates a new builder with default values. */
    public static Builder builder() {
        return new Builder();
    }

    /** Returns a configuration with all default values. */
    public static CheckConfig defaults() {
        return builder().build();
    }

    /** Builder for {@link CheckConfig}. */
    public static final class Builder {
        private Duration interval = DEFAULT_INTERVAL;
        private Duration timeout = DEFAULT_TIMEOUT;
        private Duration initialDelay = DEFAULT_INITIAL_DELAY;
        private int failureThreshold = DEFAULT_FAILURE_THRESHOLD;
        private int successThreshold = DEFAULT_SUCCESS_THRESHOLD;

        private Builder() {}

        /** Sets the check interval. */
        public Builder interval(Duration interval) {
            this.interval = interval;
            return this;
        }

        /** Sets the check timeout (must be less than interval). */
        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }

        /** Sets the initial delay before the first check. */
        public Builder initialDelay(Duration initialDelay) {
            this.initialDelay = initialDelay;
            return this;
        }

        /** Sets the consecutive failure threshold. */
        public Builder failureThreshold(int failureThreshold) {
            this.failureThreshold = failureThreshold;
            return this;
        }

        /** Sets the consecutive success threshold. */
        public Builder successThreshold(int successThreshold) {
            this.successThreshold = successThreshold;
            return this;
        }

        /** Builds and validates the configuration. */
        public CheckConfig build() {
            validate();
            return new CheckConfig(this);
        }

        private void validate() {
            if (interval.compareTo(MIN_INTERVAL) < 0 || interval.compareTo(MAX_INTERVAL) > 0) {
                throw new ValidationException(
                        "interval must be between " + MIN_INTERVAL + " and " + MAX_INTERVAL
                                + ", got " + interval);
            }
            if (timeout.compareTo(MIN_TIMEOUT) < 0 || timeout.compareTo(MAX_TIMEOUT) > 0) {
                throw new ValidationException(
                        "timeout must be between " + MIN_TIMEOUT + " and " + MAX_TIMEOUT
                                + ", got " + timeout);
            }
            if (timeout.compareTo(interval) >= 0) {
                throw new ValidationException(
                        "timeout (" + timeout + ") must be less than interval (" + interval + ")");
            }
            if (initialDelay.compareTo(MIN_INITIAL_DELAY) < 0
                    || initialDelay.compareTo(MAX_INITIAL_DELAY) > 0) {
                throw new ValidationException(
                        "initialDelay must be between " + MIN_INITIAL_DELAY + " and "
                                + MAX_INITIAL_DELAY + ", got " + initialDelay);
            }
            if (failureThreshold < MIN_THRESHOLD || failureThreshold > MAX_THRESHOLD) {
                throw new ValidationException(
                        "failureThreshold must be between " + MIN_THRESHOLD + " and "
                                + MAX_THRESHOLD + ", got " + failureThreshold);
            }
            if (successThreshold < MIN_THRESHOLD || successThreshold > MAX_THRESHOLD) {
                throw new ValidationException(
                        "successThreshold must be between " + MIN_THRESHOLD + " and "
                                + MAX_THRESHOLD + ", got " + successThreshold);
            }
        }
    }
}

File: sdk-csharp/DepHealth.Core/CheckConfig.cs
namespace DepHealth;

/// <summary>
/// Dependency health check configuration. Immutable, created via Builder.
/// </summary>
public sealed class CheckConfig
{
    /// <summary>Default value for <see cref="Interval"/>.</summary>
    public static readonly TimeSpan DefaultInterval = TimeSpan.FromSeconds(15);

    /// <summary>Default value for <see cref="Timeout"/>.</summary>
    public static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(5);

    /// <summary>Default value for <see cref="InitialDelay"/>.</summary>
    public static readonly TimeSpan DefaultInitialDelay = TimeSpan.FromSeconds(5);

    /// <summary>Default value for <see cref="FailureThreshold"/>.</summary>
    public const int DefaultFailureThreshold = 1;

    /// <summary>Default value for <see cref="SuccessThreshold"/>.</summary>
    public const int DefaultSuccessThreshold = 1;

    /// <summary>Minimum allowed value for <see cref="Interval"/>.</summary>
    public static readonly TimeSpan MinInterval = TimeSpan.FromSeconds(1);

    /// <summary>Maximum allowed value for <see cref="Interval"/>.</summary>
    public static readonly TimeSpan MaxInterval = TimeSpan.FromMinutes(10);

    /// <summary>Minimum allowed value for <see cref="Timeout"/>.</summary>
    public static readonly TimeSpan MinTimeout = TimeSpan.FromMilliseconds(100);

    /// <summary>Maximum allowed value for <see cref="Timeout"/>.</summary>
    public static readonly TimeSpan MaxTimeout = TimeSpan.FromSeconds(30);

    /// <summary>Minimum allowed value for <see cref="InitialDelay"/>.</summary>
    public static readonly TimeSpan MinInitialDelay = TimeSpan.Zero;

    /// <summary>Maximum allowed value for <see cref="InitialDelay"/>.</summary>
    public static readonly TimeSpan MaxInitialDelay = TimeSpan.FromMinutes(5);

    /// <summary>Minimum allowed threshold value.</summary>
    public const int MinThreshold = 1;

    /// <summary>Maximum allowed threshold value.</summary>
    public const int MaxThreshold = 10;

    /// <summary>Interval between health checks (default 15s).</summary>
    public TimeSpan Interval { get; }

    /// <summary>Timeout for a single health check (default 5s).</summary>
    public TimeSpan Timeout { get; }

    /// <summary>Delay before the first health check (default 5s).</summary>
    public TimeSpan InitialDelay { get; }

    /// <summary>Number of consecutive failures before marking unhealthy (default 1).</summary>
    public int FailureThreshold { get; }

    /// <summary>Number of consecutive successes before marking healthy (default 1).</summary>
    public int SuccessThreshold { get; }

    private CheckConfig(Builder builder)
    {
        Interval = builder.IntervalValue;
        Timeout = builder.TimeoutValue;
        InitialDelay = builder.InitialDelayValue;
        FailureThreshold = builder.FailureThresholdValue;
        SuccessThreshold = builder.SuccessThresholdValue;
    }

    /// <summary>Creates a new builder for constructing a <see cref="CheckConfig"/>.</summary>
    public static Builder CreateBuilder() => new();

    /// <summary>Creates a <see cref="CheckConfig"/> with default values.</summary>
    public static CheckConfig Defaults() => CreateBuilder().Build();

    /// <summary>
    /// Fluent builder for constructing a <see cref="CheckConfig"/> instance.
    /// </summary>
    public sealed class Builder
    {
        internal TimeSpan IntervalValue = DefaultInterval;
        internal TimeSpan TimeoutValue = DefaultTimeout;
        internal TimeSpan InitialDelayValue = DefaultInitialDelay;
        internal int FailureThresholdValue = DefaultFailureThreshold;
        internal int SuccessThresholdValue = DefaultSuccessThreshold;

        internal Builder() { }

        /// <summary>Sets the interval between health checks.</summary>
        /// <param name="interval">Interval (1s–10min).</param>
        public Builder WithInterval(TimeSpan interval)
        {
            IntervalValue = interval;
            return this;
        }

        /// <summary>Sets the timeout for a single health check.</summary>
        /// <param name="timeout">Timeout (100ms–30s, must be less than interval).</param>
        public Builder WithTimeout(TimeSpan timeout)
        {
            TimeoutValue = timeout;
            return this;
        }

        /// <summary>Sets the delay before the first health check.</summary>
        /// <param name="initialDelay">Initial delay (0–5min).</param>
        public Builder WithInitialDelay(TimeSpan initialDelay)
        {
            InitialDelayValue = initialDelay;
            return this;
        }

        /// <summary>Sets the number of consecutive failures before marking unhealthy.</summary>
        /// <param name="failureThreshold">Failure threshold (1–10).</param>
        public Builder WithFailureThreshold(int failureThreshold)
        {
            FailureThresholdValue = failureThreshold;
            return this;
        }

        /// <summary>Sets the number of consecutive successes before marking healthy.</summary>
        /// <param name="successThreshold">Success threshold (1–10).</param>
        public Builder WithSuccessThreshold(int successThreshold)
        {
            SuccessThresholdValue = successThreshold;
            return this;
        }

        /// <summary>Validates and builds the <see cref="CheckConfig"/> instance.</summary>
        /// <exception cref="ValidationException">Thrown when validation fails.</exception>
        public CheckConfig Build()
        {
            Validate();
            return new CheckConfig(this);
        }

        private void Validate()
        {
            if (IntervalValue < MinInterval || IntervalValue > MaxInterval)
            {
                throw new ValidationException(
                    $"interval must be between {MinInterval} and {MaxInterval}, got {IntervalValue}");
            }

            if (TimeoutValue < MinTimeout || TimeoutValue > MaxTimeout)
            {
                throw new ValidationException(
                    $"timeout must be between {MinTimeout} and {MaxTimeout}, got {TimeoutValue}");
            }

            if (TimeoutValue >= IntervalValue)
            {
                throw new ValidationException(
                    $"timeout ({TimeoutValue}) must be less than interval ({IntervalValue})");
            }

            if (InitialDelayValue < MinInitialDelay || InitialDelayValue > MaxInitialDelay)
            {
                throw new ValidationException(
                    $"initialDelay must be between {MinInitialDelay} and {MaxInitialDelay}, got {InitialDelayValue}");
            }

            if (FailureThresholdValue < MinThreshold || FailureThresholdValue > MaxThreshold)
            {
                throw new ValidationException(
                    $"failureThreshold must be between {MinThreshold} and {MaxThreshold}, got {FailureThresholdValue}");
            }

            if (SuccessThresholdValue < MinThreshold || SuccessThresholdValue > MaxThreshold)
            {
                throw new ValidationException(
                    $"successThreshold must be between {MinThreshold} and {MaxThreshold}, got {SuccessThresholdValue}");
            }
        }
    }
}
